IPython/core/completer.py

```diff
@@ -417,6 +537,189 @@ def __hash__(self):
+
+    #: Suffix of the provided ``CompletionContext.token``, if not given defaults to full token.
+    matched_fragment: NotRequired[str]
+
+    #: Whether to suppress results from all other matchers (True), some
+    #: matchers (set of identifiers) or none (False); default is False.
+    suppress: NotRequired[Union[bool, Set[str]]]
+
+    #: Identifiers of matchers which should NOT be suppressed when this matcher
+    #: requests to suppress all other matchers; defaults to an empty set.
+    do_not_suppress: NotRequired[Set[str]]
+
+    #: Are completions already ordered and should be left as-is? default is False.
+    ordered: NotRequired[bool]
+
+
+@sphinx_options(show_inherited_members=True, exclude_inherited_from=["dict"])
+class SimpleMatcherResult(_MatcherResultBase, TypedDict):
+    """Result of new-style completion matcher."""
+
+    # note: TypedDict is added again to the inheritance chain
+    # in order to get __orig_bases__ for documentation
+
+    #: List of candidate completions
+    completions: Sequence[SimpleCompletion]
+
+
+class _JediMatcherResult(_MatcherResultBase):
+    """Matching result returned by Jedi (will be processed differently)"""
+
+    #: list of candidate completions
+    completions: Iterable[_JediCompletionLike]
+
+
+class CompletionContext(NamedTuple):
+    """Completion context provided as an argument to matchers in the Matcher API v2."""
+
+    # rationale: many legacy matchers relied on completer state (`self.text_until_cursor`)
+    # which was not explicitly visible as an argument of the matcher, making any refactor
+    # prone to errors; by explicitly passing `cursor_position` we can decouple the matchers
+    # from the completer, and make substituting them in sub-classes easier.
+
+    #: Relevant fragment of code directly preceding the cursor.
+    #: The extraction of token is implemented via splitter heuristic
+    #: (following readline behaviour for legacy reasons), which is user configurable
+    #: (by switching the greedy mode).
+    token: str
+
+    #: The full available content of the editor or buffer
+    full_text: str
+
+    #: Cursor position in the line (the same for ``full_text`` and ``text``).
+    cursor_position: int
+
+    #: Cursor line in ``full_text``.
+    cursor_line: int
+
+    #: The maximum number of completions that will be used downstream.
+    #: Matchers can use this information to abort early.
+    #: The built-in Jedi matcher is currently excepted from this limit.
+    # If not given, return all possible completions.
+    limit: Optional[int]
+
+    @property
+    @lru_cache(maxsize=None)  # TODO change to @cache after dropping Python 3.7
+    def text_until_cursor(self) -> str:
+        return self.line_with_cursor[: self.cursor_position]
+
+    @property
+    @lru_cache(maxsize=None)  # TODO change to @cache after dropping Python 3.7
+    def line_with_cursor(self) -> str:
+        return self.full_text.split("\n")[self.cursor_line]
```

Comment:
We are already 3.8+ https://github.com/ipython/ipython/blob/main/setup.cfg#L29, I think you mean after dropping Python 3.8. 

One easy thing is at the top of the file you can do 

```
if sys.version_info < (3, 8):
   from functools import lru_cahce
   cache = lru_cache(maxsize=None)
else:
   from functools import cache
```

I grep from `sys_version_info` every time we drop a Python version.

Reply:
Yes, it should have been < 3.8. In ab60d3197788b943db0cf81d9279c28e52ec9225 I switched to `@cached_property` instead as using `@property @cache` as it was a premature optimisation (it would be more memory efficient as per documentation, but at any time there is only one instance of `CompletionContext`, so there is no need to worry about key-sharing optimisation - and for the same reason we can switch it to `dataclass` which has a benefit of narrower API)